### ISSUE TITLE
Feature: Variable/Weighted ChunkedDribbleDelay

### ISSUE THREAD

**Issue by tjcelaya** [tomakehurst/wiremock#966]
**TL;DR**: how would I achieve one of the following to simulate more real-world network failure:
- inserting delays into _just the response body_ (or at other points in the payload, be it through byte offsets, percentages, etc.)
- chunked dribble delay but weighted towards the end of the response so that the timeout is most likely to occur while downloading the response body

Long explanation and sample code follows:

I was experimenting with wiremock and trying to simulate a fault during the HTTP response body. Unfortunately, I haven't found a combination of client socket timeout and `withChunkedDribbleDelay` that produces the intended behavior. Regardless of the values provided, it appears to be impossible to force the delays to occur during the download of the response body but not the headers.

Assuming that this would be possible with extensions I also implemented a `ResponseTransformer` that would build a new `Response` by copying over all the fields that are passed through the constructor and wrapping the source stream in a custom `InputStream` which keeps an internal counter and sleeps longer during every call to the various `read()` methods. Unfortunately this doesn't seem to work in combination with proxying because the entire response is read into a byte array and the ResponseTransformer seems to affect how the proxy reads from the downstream instead of how the response is read from the proxy. (I'm not sure if this would work with a non-proxy response, haven't had a chance to test that).

I haven't been able to figure out a way to offset where the delays are inserted without extending `WireMockHandlerDispatchingServlet` and all the intermediate classes (`JettyHttpServerFactory`/`JettyHttpServer`/etc).

`ResponseTransformer` for `IncreasingDelayReadingBodyResponse`/`IncreasingDelayInputStream`:

```java

    private static class IncreasingDelayInputStream extends ProxyInputStream {

        private static final Logger LOG = LoggerFactory.getLogger(IncreasingDelayInputStream.class);
        
        private final int baseDelay;

        private final float mult;

        private int n;

        IncreasingDelayInputStream(final InputStream proxy,
                                   final int baseDelay,
                                   final float mult) {
            super(proxy);
            this.baseDelay = baseDelay;
            this.mult = mult;
            this.n = 0;
        }

        @Override
        protected void beforeRead(final int n) {
            try {
                final int sleeping = round(this.n++ * this.baseDelay * this.mult);
                LOG.info(" > sleeping " + sleeping);
                Thread.sleep(sleeping);
            } catch (final InterruptedException ie) {
                Thread.currentThread().interrupt();
                throw new RuntimeException(ie);
            }
        }
    }

    private static class IncreasingDelayReadingBodyResponse extends Response {
        IncreasingDelayReadingBodyResponse(final Response response,
                                           final int baseDelay,
                                           final float mult) {
            super(response.getStatus(),
                  response.getStatusMessage(),
                  () -> new IncreasingDelayInputStream(response.getBodyStream(), baseDelay, mult),
                  response.getHeaders(),
                  response.wasConfigured(),
                  response.getFault(),
                  response.getInitialDelay(),
                  response.getChunkedDribbleDelay(),
                  response.isFromProxy());
        }
    }

    public void setup() {
        final WireMockConfiguration wmc = wireMockConfig()
                .dynamicPort()
                .disableRequestJournal()
                .extensions(new ResponseTransformer() {
                    @Override
                    public Response transform(final Request request,
                                              final Response response,
                                              final FileSource files,
                                              final Parameters parameters) {
                        return new IncreasingDelayReadingBodyResponse(response, 200, 1.1f);
                    }

                    @Override
                    public String getName() {
                        return "increasing-delay";
                    }
                });
        // more non-wiremock configuration...
    }

    public void testthings() {
         // HttpClient configuration...
         final CloseableHttpClient client = httpClientBuilder.build();

            // allocate small buffer for reading to make delays more obvious
            try (final CloseableHttpResponse response = client.execute(get)) {
                final HttpEntity entity = response.getEntity();
                    try (final InputStream body = entity.getContent()) {
                        // read bytes into small buffer, print, timing, continue
                    }
    }
```

Log output with IncreasingDelayInputStream only:

```
// this first copy seems to be when the Response is wrapped in a
// LoggedResponse and #getBody() is called
[qtp838449246-14] INFO  IncreasingDelayInputStream [ ] -  > sleeping 0
[qtp838449246-14] INFO  IncreasingDelayInputStream [ ] -  > sleeping 220
// this copy occurs in WireMockHandlerDispatchingServlet#writeAndTranslateExceptions
// which is where I'd assume the delays from IncreasingDelayInputStream
[qtp838449246-14] INFO  IncreasingDelayInputStream [ ] -  > sleeping 0
[qtp838449246-14] INFO  IncreasingDelayInputStream [ ] -  > sleeping 220
[main] WARN  HttpClientIT [ ] -  >> (1069 ms) >> read header
[main] INFO  HttpClientIT [ ] - response line: HTTP/1.1 200 OK
[main] INFO  HttpClientIT [ ] - response body (streamed):
[main] WARN  HttpClientIT [ ] -  >> (  11 ms) >> {"args":{"foo":"
[main] WARN  HttpClientIT [ ] -  >> (   0 ms) >> bar"},"headers":
[main] WARN  HttpClientIT [ ] -  >> (   0 ms) >> {"host":"postman
[main] WARN  HttpClientIT [ ] -  >> (   0 ms) >> -echo.com","acce
[main] WARN  HttpClientIT [ ] -  >> (   0 ms) >> pt-encoding":"gz
[main] WARN  HttpClientIT [ ] -  >> (   0 ms) >> ip,deflate","use
[main] WARN  HttpClientIT [ ] -  >> (   0 ms) >> r-agent":"Apache
[main] WARN  HttpClientIT [ ] -  >> (   0 ms) >> -HttpClient/4.5.
[main] WARN  HttpClientIT [ ] -  >> (   0 ms) >> 5 (Java/1.8.0_13
[main] WARN  HttpClientIT [ ] -  >> (   0 ms) >> 1)","x-forwarded
[main] WARN  HttpClientIT [ ] -  >> (   0 ms) >> -port":"443","x-
[main] WARN  HttpClientIT [ ] -  >> (   0 ms) >> forwarded-proto"
[main] WARN  HttpClientIT [ ] -  >> (   0 ms) >> :"https"},"url":
[main] WARN  HttpClientIT [ ] -  >> (   0 ms) >> "https://postman
[main] WARN  HttpClientIT [ ] -  >> (   0 ms) >> -echo.com/get/?f
[main] WARN  HttpClientIT [ ] -  >> (   0 ms) >> oo=bar"}
[main] WARN  HttpClientIT [ ] - runtime: 1094 ms
```

Log output with ChunkedDribbleDelay only:

```
[main] WARN  HttpClientIT [ ] -  >> (1253 ms) >> read header
[main] INFO  HttpClientIT [ ] - response line: HTTP/1.1 200 OK
[main] INFO  HttpClientIT [ ] - response body (streamed):
[main] WARN  HttpClientIT [ ] -  >> ( 301 ms) >> {"args":{"foo":"
[main] WARN  HttpClientIT [ ] -  >> (   0 ms) >> bar"},"headers":
[main] WARN  HttpClientIT [ ] -  >> (  99 ms) >> {"host":"postman
[main] WARN  HttpClientIT [ ] -  >> (   0 ms) >> -echo.com","acce
[main] WARN  HttpClientIT [ ] -  >> ( 102 ms) >> pt-encoding":"gz
[main] WARN  HttpClientIT [ ] -  >> (   0 ms) >> ip,deflate","use
[main] WARN  HttpClientIT [ ] -  >> ( 100 ms) >> r-agent":"Apache
[main] WARN  HttpClientIT [ ] -  >> (   0 ms) >> -HttpClient/4.5.
[main] WARN  HttpClientIT [ ] -  >> ( 104 ms) >> 5 (Java/1.8.0_13
[main] WARN  HttpClientIT [ ] -  >> (   0 ms) >> 1)","x-forwarded
[main] WARN  HttpClientIT [ ] -  >> ( 103 ms) >> -port":"443","x-
[main] WARN  HttpClientIT [ ] -  >> (   0 ms) >> forwarded-proto"
[main] WARN  HttpClientIT [ ] -  >> (   0 ms) >> :"https"},"url":
[main] WARN  HttpClientIT [ ] -  >> ( 104 ms) >> "https://postman
[main] WARN  HttpClientIT [ ] -  >> (   0 ms) >> -echo.com/get/?f
[main] WARN  HttpClientIT [ ] -  >> (   0 ms) >> oo=bar"}
[main] WARN  HttpClientIT [ ] - runtime: 2178 ms
```

**Comment by dekobon** [comment]
:+1:

**Comment by oberlies** [comment]
This feature is also what I am looking for: The possibility to only delay the response body but not the response headers. (BTW I would have expected that "chunked dribble delay" would do exactly this because only the response body is transferred as chunks; the current implementation is rather a "frame dribble delay"...)

My use case is to test timeouts in a reactive client. Unfortunately I had a bug in the timeout handling because I effectively only specified a timeout for the response headers, and none for the body. So good test support is urgently needed here!
